lib/cache/memory-cache-store.js

```diff
@@ -0,0 +1,402 @@
+
+  /**
+   * @type {Map<string, Map<string, MemoryStoreValue>>}
+   */
+  #data = new Map()
+
+  /**
+   * @param {import('../../types/cache-interceptor.d.ts').default.MemoryCacheStoreOpts | undefined} opts
+   */
+  constructor (opts) {
+    if (opts) {
+      if (typeof opts !== 'object') {
+        throw new TypeError('MemoryCacheStore options must be an object')
+      }
+
+      if (opts.maxEntries !== undefined) {
+        if (
+          typeof opts.maxEntries !== 'number' ||
+          !Number.isInteger(opts.maxEntries) ||
+          opts.maxEntries < 0
+        ) {
+          throw new TypeError('MemoryCacheStore options.maxEntries must be a non-negative integer')
+        }
+        this.#maxEntries = opts.maxEntries
+      }
+
+      if (opts.maxEntrySize !== undefined) {
+        if (
+          typeof opts.maxEntrySize !== 'number' ||
+          !Number.isInteger(opts.maxEntrySize) ||
+          opts.maxEntrySize < 0
+        ) {
+          throw new TypeError('MemoryCacheStore options.maxEntrySize must be a non-negative integer')
+        }
+        this.#maxEntrySize = opts.maxEntrySize
+      }
+    }
+  }
+
+  get isFull () {
+    return this.#entryCount >= this.#maxEntries
+  }
+
+  /**
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @returns {import('../../types/cache-interceptor.d.ts').default.CacheStoreReadable | undefined}
+   */
+  createReadStream (req) {
+    if (typeof req !== 'object') {
+      throw new TypeError(`expected req to be object, got ${typeof req}`)
+    }
+
+    const values = this.#getValuesForRequest(req, false)
+    if (!values) {
+      return undefined
+    }
+
+    const value = this.#findValue(req, values)
+
+    if (!value || value.readLock) {
+      return undefined
+    }
+
+    return new MemoryStoreReadableStream(value)
+  }
+
+  /**
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @param {import('../../types/cache-interceptor.d.ts').default.CacheStoreValue} opts
+   * @returns {import('../../types/cache-interceptor.d.ts').default.CacheStoreWriteable | undefined}
+   */
+  createWriteStream (req, opts) {
+    if (typeof req !== 'object') {
+      throw new TypeError(`expected req to be object, got ${typeof req}`)
+    }
+    if (typeof opts !== 'object') {
+      throw new TypeError(`expected value to be object, got ${typeof opts}`)
+    }
+
+    if (this.isFull) {
+      return undefined
+    }
+
+    const values = this.#getValuesForRequest(req, true)
+
+    let value = this.#findValue(req, values)
+    if (!value) {
+      // The value doesn't already exist, meaning we haven't cached this
+      //  response before. Let's assign it a value and insert it into our data
+      //  property.
+
+      if (this.isFull) {
+        // Or not, we don't have space to add another response
+        return undefined
+      }
+
+      this.#entryCount++
+
+      value = {
+        readers: 0,
+        readLock: false,
```

Comment:
Can you repeat/clarify why are using read locks for the stream?
This part I'm not fully getting it.

Reply:
The readLock is to stop a request from writing to a value if there's another request reading from it.

The writeLock is to stop a request from reading a value that's currently being written to

Reply:
We must iterate over this, but it's fine for now